### PR TITLE
Hypre runtime compute policy

### PR DESCRIPTION
**Pull request by tomstitt**
This PR adds support for HYPRE's runtime compute/memory policy. HYPRE's runtime policy requires at least HYPRE 2.26.0 and the lazy device library initialization requires https://github.com/hypre-space/hypre/pull/962, which ~is not yet in a release~ requires HYPRE 2.31.0. ~We will need to change MFEM examples if we want to setup HYPRE in `Hypre::Init()` because all examples I've seen call `Hypre::Init()` before setting up the `mfem::Device` but there are now checks under `Hypre::Init()` that use the `mfem::Device` state.~

~Thinking about this more it may make sense to add a configuration option to enable this so that we can merge without changing existing behavior and so that we don't break people with `Hypre::Init` calls before `Device` setup like in the examples. Thoughts?~

todo:
- [x] guard more with `#if MFEM_HYPRE_VERSION >= 22600`/backward compatibility
~- [ ] update examples/tests/miniapps using `HYPRE_USING_GPU`~

~or~

~- [ ] make these changes optional and check for something like `MEFM_HYPRE_USE_RUNTIME_POLICY` instead of `MFEM_HYPRE_VERSION >= 22600`~

Update: Motivated by some discussion with @v-dobrev  I moved the HYPRE compute/memory policy init to its own method and call that from `Device::Configure`; by default HYPRE's policy with match MFEM. I confirmed this is doing what we want by running `rocprof` on `ex1p` with `-d cpu` and `-d hip` and see zero device activity in the former and HYPRE & MFEM kernel activity in the later. If a user wants to ignore MFEM's policy and use the HYPRE default they can set `mfem::Hypre::configure_hypre_runtime_policy_from_mfem = false` before initializing their `Device`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3844](https://github.com/mfem/mfem/pull/3844) | @tomstitt | @tzanio | @stefanozampini + @v-dobrev | 11/29/23 | 5/4/24 | 5/4/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
@tomstitt, do you think this will be ready in time for [mfem-4.6](https://github.com/mfem/mfem/pull/3841)?

**Comment by tomstitt** [comment]
> @tomstitt, do you think this will be ready in time for [mfem-4.6](https://github.com/mfem/mfem/pull/3841)?

Should be, I'm almost done testing locally. Next I need to cleanup what I have a little and fix backward compatibility and go through the examples/tests/miniapps and handle the `HYPRE_USING_GPU` blocks they have

**Comment by tzanio** [comment]
@tomstitt, is this ready for review?

**Comment by tomstitt** [comment]
> @tomstitt, is this ready for review?

No.. not yet. It's taking me longer to get full backwards compatibility than expected. We also need https://github.com/hypre-space/hypre/tree/gpu_init, which isn't merged yet but I've reached out to try to get an ETA.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by sebastiangrimberg** [comment]
Update: This functionality is now in a Hypre release version (2.31, after https://github.com/hypre-space/hypre/pull/1068).

**Comment by tzanio** [comment]
Now that PR https://github.com/mfem/mfem/pull/3922 has been merged in `master`, please update this PR by:

1. merging `master` in here and resolving conflicts
2. replacing use of `double` with `real_t`
3. guarding double/single-only sections with `MFEM_USE_DOUBLE` and `MFEM_USE_SINGLE`
4. making sure tests run cleanly with default and `MFEM_PRECISION=single` configurations

See https://github.com/orgs/mfem/discussions/4207 for more details and/or ask @dylan-copeland for help.

File: linalg/hypre.hpp [hunk 1]
```diff
@@ -1036,18 +1049,18 @@ class HypreSmoother : public Solver
...
+
+   Type DefaultType()
+   {
+      return HypreUsingGPU() ? l1Jacobi : l1GS;
+   }
 
    HypreSmoother();
 
-   HypreSmoother(const HypreParMatrix &A_, int type = default_type,
+   HypreSmoother(const HypreParMatrix &A_, int type = Undefined,
```
[the remark this answers is not in the document]
Comment (tomstitt) [reply linalg/hypre.hpp:1071]: good call!! I didn't realize you could do that but I see that default arguments are actually evaluated each time the ctor is called

File: linalg/hypre_parcsr.hpp [hunk 2]
```diff
@@ -29,6 +29,34 @@ typedef HYPRE_Int HYPRE_BigInt;
 #define HYPRE_MPI_BIG_INT HYPRE_MPI_INT
 #endif
 
+
+namespace mfem
+{
```
[the remark this answers is not in the document]
Comment (tomstitt) [reply linalg/hypre_parcsr.hpp:34]: yes !

**Comment by tzanio** [comment]
@v-dobrev and @sebastiangrimberg -- This has been updated so the examples do not require any changes. Can you take another look?

@tomstitt -- Can you add an item about this in `CHANGELOG`?

**Comment by sebastiangrimberg** [comment]
> @v-dobrev and @sebastiangrimberg -- This has been updated so the examples do not require any changes. Can you take another look?

I think this approach looks great. I haven't tested it on a GPU build, but can next week if needed. The changes look good to me to merge as is.

**Comment by tomstitt** [comment]
> @tomstitt -- Can you add an item about this in `CHANGELOG`?

Done!

**Comment by tzanio** [comment]
Thanks @tomstitt !

**Comment by sebastiangrimberg** [comment]
Sorry for the delay here, I just got around to building and testing with CUDA support. I think this looks good, I just have one suggested style change to make the unit test consistent with the examples in their initialization of initialize Hypre, followed by the `mfem::Device` (this doesn't change functionality by the design):

```
diff --git a/tests/unit/pcunit_test_main.cpp b/tests/unit/pcunit_test_main.cpp
index e45f495d8..0ba491bf9 100644
--- a/tests/unit/pcunit_test_main.cpp
+++ b/tests/unit/pcunit_test_main.cpp
@@ -26,11 +26,11 @@ int main(int argc, char *argv[])
    return MFEM_SKIP_RETURN_VALUE;
 #endif

-   mfem::Device device("cuda");
 #ifdef MFEM_USE_MPI
    mfem::Mpi::Init();
    mfem::Hypre::Init();
 #endif
+   mfem::Device device("cuda");

    // Include only tests that are labeled with both CUDA and Parallel.
    return RunCatchSession(argc, argv, {"[CUDA]","[Parallel]"}, Root());
```

and

```
diff --git a/tests/unit/cunit_test_main.cpp b/tests/unit/cunit_test_main.cpp
index 2c5829d19..9e31f1b34 100644
--- a/tests/unit/cunit_test_main.cpp
+++ b/tests/unit/cunit_test_main.cpp
@@ -22,6 +22,7 @@ int main(int argc, char *argv[])
 #endif

    mfem::Device device("cuda");
+
    // Include only tests labeled with CUDA. Exclude parallel tests.
    return RunCatchSession(argc, argv, {"[CUDA]", "~[Parallel]"});
 }
```

I am also getting the inverse of the behavior I'd expect for the unit tests, but correct me if I'm wrong.

- Running `pcunit_tests` with CUDA builds of MFEM and Hypre produces the same results on both `master` and this PR (though I am getting a test failure for `"Parallel H1 Full Assembly"` where the results are not equal, not sure why this is. Maybe something to do with the version of Hypre I am using or something like that?)
- Running `punit_tests` works fine on `master` with GPU builds, but on this PR I get the following error:

```
Verification failed: (!alias) is false:
 --> Cannot register an alias!
 ... in function: static void* mfem::MemoryManager::Register_(void*, void*, size_t, mfem::MemoryType, bool, bool, unsigned int&)
 ... in file: /software/palace/build-cuda-x86_64/extern/mfem/general/mem_manager.cpp:826
```

@tomstitt, can you reproduce this? I would have expected the `punit_tests` test to fail on `master` since MFEM would be using CPU and Hypre expecting GPU, but this PR should support that.

**Comment by sebastiangrimberg** [comment]
Here's the diff to fix running `unit_tests` and `punit_tests` for GPU builds of Hypre:

```
diff --git a/tests/unit/punit_test_main.cpp b/tests/unit/punit_test_main.cpp
index ecc45fcb7..cda8cd2ed 100644
--- a/tests/unit/punit_test_main.cpp
+++ b/tests/unit/punit_test_main.cpp
@@ -30,6 +30,7 @@ int main(int argc, char *argv[])
    mfem::Mpi::Init();
    mfem::Hypre::Init();
 #endif
+   mfem::Device device("cpu");

    // Only run tests that are labeled with Parallel.
    return RunCatchSession(argc, argv, {"[Parallel]"}, Root());
diff --git a/tests/unit/unit_test_main.cpp b/tests/unit/unit_test_main.cpp
index af6f6bb98..f8a550dd2 100644
--- a/tests/unit/unit_test_main.cpp
+++ b/tests/unit/unit_test_main.cpp
@@ -21,6 +21,8 @@ int main(int argc, char *argv[])
    return MFEM_SKIP_RETURN_VALUE;
 #endif

+   mfem::Device device("cpu");
+
    // Exclude parallel tests.
    return RunCatchSession(argc, argv, {"~[Parallel]"});
 }
```

This ensures `Hypre::InitDevice` gets called which correctly tells Hypre to not use the GPU based on the `mfem::Device` configuration (otherwise we get the error above about aliasing).

I also found a few issues with unit tests for GPU builds when not using the device. Those can be found in 1fbeee2.

**Comment by tzanio** [comment]
thanks @sebastiangrimberg !

**Comment by tomstitt** [comment]
Thank you @sebastiangrimberg for fixing the unit tests! I completely neglected those =|

> I am also getting the inverse of the behavior I'd expect for the unit tests, but correct me if I'm wrong.
> 
> * Running `pcunit_tests` with CUDA builds of MFEM and Hypre produces the same results on both `master` and this PR (though I am getting a test failure for `"Parallel H1 Full Assembly"` where the results are not equal, not sure why this is. Maybe something to do with the version of Hypre I am using or something like that?)
> * Running `punit_tests` works fine on `master` with GPU builds, but on this PR I get the following error:
> 
> ```
> Verification failed: (!alias) is false:
>  --> Cannot register an alias!
>  ... in function: static void* mfem::MemoryManager::Register_(void*, void*, size_t, mfem::MemoryType, bool, bool, unsigned int&)
>  ... in file: /software/palace/build-cuda-x86_64/extern/mfem/general/mem_manager.cpp:826
> ```
> 
> @tomstitt, can you reproduce this? I would have expected the `punit_tests` test to fail on `master` since MFEM would be using CPU and Hypre expecting GPU, but this PR should support that.

I see both errors and they are unexpected to me, let me take a look and try out your fixes !

**Comment by v-dobrev** [comment]
⚠️ There are some changes in `master` that will probably need to be updated for the runtime Hypre device configuration. 

One example is this: https://github.com/mfem/mfem/blob/756fd52c2b078acd4c64e235b5fb24093fb30d83/linalg/hypre.cpp#L3576-L3603

**Review by sebastiangrimberg — APPROVED** [review]
This looks good to me now. I think are is a missing newline needed in the `CHANGELOG` but figure that will be resolved by the final check in https://github.com/mfem/mfem/pull/4172#issue-2167467001.

**Comment by v-dobrev** [comment]
I think @sebastiangrimberg's suggestions for the unit tests make sense and I added them, except the one for `unit_test_main.cpp` where adding
```c++
   mfem::Device device("cpu");
```
as suggested, will not make a difference, I think, since this test will not use hypre even when using a parallel MFEM build.

**Comment by sebastiangrimberg** [comment]
The CI test failure is the result of the `sedov` tests reusing the `main` function from `punit_test_main.cpp` in their binary and thus configuring the `Device` twice: First in `punit_test_main.cpp` and then again in `test_sedov.cpp` (in the `TEST_CASE("Sedov")` block). The first solution that comes to mind is to move the Sedov device configuration into the respective `main` functions so no unit tests are configuring the `Device` inside of a `TEST_CASE`.

**Comment by sebastiangrimberg** [comment]
I have pushed changes to fix the miniapps unit tests to a branch [`sjg/hypre-runtime-compute-policy`](https://github.com/mfem/mfem/compare/hypre-runtime-compute-policy...sjg/hypre-runtime-compute-policy). Rather than [the above suggestion](https://github.com/mfem/mfem/pull/3844#issuecomment-2093435801), it was easiest to just have `test_sedov.cpp` and `test_tmop_pa.cpp` just define their own `main` functions, as is done in `test_debug_device.cpp`.

**Comment by v-dobrev** [comment]
> I have pushed changes to fix the miniapps unit tests to a branch [`sjg/hypre-runtime-compute-policy`](https://github.com/mfem/mfem/compare/hypre-runtime-compute-policy...sjg/hypre-runtime-compute-policy). Rather than [the above suggestion](https://github.com/mfem/mfem/pull/3844#issuecomment-2093435801), it was easiest to just have `test_sedov.cpp` and `test_tmop_pa.cpp` just define their own `main` functions, as is done in `test_debug_device.cpp`.
> 
> If these look good, feel free to merge them here @v-dobrev. They should fix both CMake and GNU Make builds.

Thanks @sebastiangrimberg! I pushed a small fix and some tweaks to your branch and then merged it here.

**Review by v-dobrev — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**